### PR TITLE
Adds modules to begin work on MicroShift

### PR DESCRIPTION
**Pull request by stevsmit**
Adds modules to begin work on MicroShift

4.12+

Based on https://github.com/openshift/microshift/blob/main/docs/devenv_rhel8.md

Issue: https://issues.redhat.com/browse/OSDOCS-4547

Pending QE/SME

Preview: https://stevsmit.github.io/openshift-docs/microshift-install-rpm.html

### CODE REVIEW

File: modules/configuring-microshift-virtual-machine.adoc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+----
+$ export GO_VER=1.18.7; curl -L -o go${GO_VER}.linux-amd64.tar.gz https://go.dev/dl/go${GO_VER}.linux-amd64.tar.gz &&
+    sudo rm -rf /usr/local/go${GO_VER} && \
+    sudo mkdir -p /usr/local/go${GO_VER} && \
+    sudo tar -C /usr/local/go${GO_VER} -xzf go${GO_VER}.linux-amd64.tar.gz --strip-components 1 && \
+    sudo rm -rfv /usr/local/bin/{go,gofmt}
+    sudo ln --symbolic /usr/local/go1.18.7/bin/go /usr/local/go1.18.7/bin/gofmt /usr/local/bin/ && \
+    rm -rfv go${GO_VER}.linux-amd64.tar.gz
+----
+
```
Comment (dhellmann) [inline modules/configuring-microshift-virtual-machine.adoc:52]:
```suggestion
```

The commands above install golang (not cockpit). Users won't need to do that step, so you could cut lines 40-52.

File: microshift_install/microshift-install-rpm.adoc [hunk 2]
```diff
@@ -6,3 +6,31 @@ include::_attributes/attributes-microshift.adoc[]
 toc::[]
 
 // The content of this file should be based on https://github.com/openshift/microshift/blob/main/docs/devenv_rhel8.md but without the steps for building from source.
+
+You can install MicroShift from an RPM on a virtual machine with {op-system-first} 8, or on a {op-system} production-similar environment. 
+
+== Installing MicroShift from RPM
+
+Use the following procedures to create a MicroShift development environment on {op-system-first} 8.7 or greater by following the content in this section.
```
Comment (dhellmann) [inline microshift_install/microshift-install-rpm.adoc:14]: We have to be careful about version numbers right now. It's only supported on 8.7, not newer versions of 8. We may move to 9 in the next release and drop support for 8.

File: modules/configuring-running-microshift-executable.adoc [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+Use the following procedure to configure and run a MicroShift deployment 
+
+.Prerequisites 
+
+* You have logged into your development virtual machine using your MicroShift user credentials. 
+* You have installed the `oc` CLI and `kubectl` utilities. 
+* You have updated CRI-O, networking, and file system customizations. 
+* You have set up your MicroShift development environment. 
+* You have configured and built MicroShift on your virtual machine.
```
Comment (dhellmann) [inline modules/configuring-running-microshift-executable.adoc:18]:
```suggestion
```

File: modules/configuring-running-microshift-executable.adoc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+Use the following procedure to configure and run a MicroShift deployment 
+
+.Prerequisites 
+
+* You have logged into your development virtual machine using your MicroShift user credentials. 
+* You have installed the `oc` CLI and `kubectl` utilities. 
+* You have updated CRI-O, networking, and file system customizations. 
+* You have set up your MicroShift development environment. 
+* You have configured and built MicroShift on your virtual machine. 
+* You have prepared the MicroShift executable runtime.
```
Comment (dhellmann) [inline modules/configuring-running-microshift-executable.adoc:19]: I'm not sure what this means. 

Users should just need to install MicroShift via the RPM we will publish.

File: modules/configuring-running-microshift-executable.adoc [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+$ sudo cat /var/lib/microshift/resources/kubeadmin/kubeconfig > ~/.kube/config
+----
+
+. You can verify that MicroShift is running by entering the following commands:
++
+[source,terminal]
+----
+$ oc get cs 
+$ watch oc get pods -A 
+----
```
Comment (dhellmann) [inline modules/configuring-running-microshift-executable.adoc:73]:
```suggestion
```

We would only want developers to follow these steps, so let's not put them in the user docs.

File: modules/creating-microshift-development-environment.adoc [hunk 6]
```diff
@@ -0,0 +1,79 @@
+// Module included in the following assemblies:
```
Comment (dhellmann) [inline modules/creating-microshift-development-environment.adoc:1]: We probably want to call this something other than the "development environment." These instructions are a good way to prepare a VM for experimenting with MicroShift, so if we can use some of that language that would be better.

File: modules/creating-microshift-development-environment.adoc [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+:_content-type: PROCEDURE
+[id="creating-microshift-development-environment"]
+= Creating the MicroShift development environment
+:context: creating-microshift-development-environment
+
+Use the following procedure to set up your MicroShift development environment. 
+
+.Procedure 
+
+. Download {op-system-first} 8.7 or greater from the link:https://developers.redhat.com/products/rhel/download[Download Red Hat Enterprise Linux] page.
```
Comment (dhellmann) [inline modules/creating-microshift-development-environment.adoc:14]:
```suggestion
. Download {op-system-first} 8.7 from the link:https://developers.redhat.com/products/rhel/download[Download Red Hat Enterprise Linux] page. 
```

File: modules/creating-microshift-development-environment.adoc [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+. Create a {op-system} 8 virtual machine with the following settings: 
+.. 2 cores
+.. 4096 MB of RAM
+.. 50 GB of storage 
++
+[NOTE]
+====
+Visual Studio Code might consume up to 2 GB of RAM. For running the IDE on the development virtual machine, it is recommended to allocate at least 6144 MB of RAM in total. 
+====
```
Comment (dhellmann) [inline modules/creating-microshift-development-environment.adoc:29]:
```suggestion
```

That note is for a MicroShift developer and we can ignore it for users.

File: modules/creating-microshift-development-environment.adoc [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+. Move the ISO image to the `/var/lib/libvirt/images` directory. 
+
+. Enter the following command to create the virtual machine:
```
Comment (dhellmann) [inline modules/creating-microshift-development-environment.adoc:33]:
```suggestion
. Enter the following commands to create the virtual machine: 
```

This step on line 33 is the implementation of the instruction for the step on line 21. Do we want to move the step on line 31 up above (the user needs to do that before creating the VM with this command) and then merge the other 2 steps into 1 thing? Or maybe the stuff about the VM parameters on lines 21-24 should move above the procedure entirely and be rephrased as something like "To experiment with MicroShift, you will need a VM with ..."

File: modules/creating-microshift-development-environment.adoc [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+. Enter the following command to create the virtual machine: 
++
+[source,terminal]
+----
+$ sudo dnf install -y libvirt virt-manager virt-viewer libvirt-client qemu-kvm qemu-img
+
+VMNAME="microshift-dev"
+sudo -b bash -c " \
+cd /var/lib/libvirt/images/ && \
+virt-install \
```
Comment (dhellmann) [inline modules/creating-microshift-development-environment.adoc:42]:
```suggestion
$ VMNAME="microshift-dev"
$ sudo -b bash -c " \
cd /var/lib/libvirt/images/ && \
virt-install \
```

There are 3 separate commands here (prefixed with the prompt `$`).

File: modules/preparing-microshift-executable.adoc [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+Use the following procedure to prepare the MicroShift executable runtime. 
+
+[NOTE]
+====
+When working with MicroShift based on a pre-release minor version `Y` of {OCP}, the corresponding RPM repository `rhocp-4.$Y-for-rhel-8-$ARCH-rpms` will not be available yet. As a workaround, use the `Y-1` version. 
+====
+
+.Prerequisites 
+
+* You have logged into your development virtual machine using your MicroShift user credentials.
```
Comment (dhellmann) [inline modules/preparing-microshift-executable.adoc:19]: When we pick another name, replace "development virtual machine" here.

File: modules/preparing-microshift-executable.adoc [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+[id="preparing-microshift-executable"]
+= Preparing the MicroShift executable runtime
+:context: creating-microshift-development-environment
+
+Use the following procedure to prepare the MicroShift executable runtime. 
+
+[NOTE]
+====
+When working with MicroShift based on a pre-release minor version `Y` of {OCP}, the corresponding RPM repository `rhocp-4.$Y-for-rhel-8-$ARCH-rpms` will not be available yet. As a workaround, use the `Y-1` version. 
+====
```
Comment (dhellmann) [inline modules/preparing-microshift-executable.adoc:15]: I'm not sure if we want to say this to customers.

@ggiguash what do you think?

File: modules/preparing-microshift-executable.adoc [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+.Prerequisites 
+
+* You have logged into your development virtual machine using your MicroShift user credentials. 
+* You have configured your MicroShift virtual machine. 
+* You have built MicroShift in your development virtual machine.
```
Comment (dhellmann) [inline modules/preparing-microshift-executable.adoc:21]:
```suggestion
```

File: modules/preparing-microshift-executable.adoc [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+* You have built MicroShift in your development virtual machine. 
+
+.Procedure
+
+. Enable the necessary repositories by entering the following command:
++
+[source,terminal]
+----
+$ sudo subscription-manager repos \
+    --enable rhocp-4.11-for-rhel-8-$(uname -i)-rpms \
```
Comment (dhellmann) [inline modules/preparing-microshift-executable.adoc:30]: @ggiguash what's the right version repo to use here? Can we predict the 4.12 repo name and use it now?

File: modules/preparing-microshift-executable.adoc [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    --enable fast-datapath-for-rhel-8-$(uname -i)-rpms
+----
+
+. Install the MicroShift RPM packages by entering the following command: 
++
+[source,terminal]
+----
+$ sudo dnf localinstall -y ~/microshift/_output/rpmbuild/RPMS/*/*.rpm
```
Comment (dhellmann) [inline modules/preparing-microshift-executable.adoc:38]:
```suggestion
$ sudo dnf install microshift
```

The command in the dev docs installs the RPMs built from source by some of the other steps. Users should install the RPMs we ship normally.

File: modules/stopping-microshift-executable.adoc [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+* You have configured and built MicroShift on your virtual machine. 
+* You have prepared the MicroShift executable runtime. 
+
+.Procedure 
+
+. Enter the following command to stop the MicroShift process: 
++
+[source,terminal]
+----
+$ sudo kill microshift && sleep 3
```
Comment (dhellmann) [inline modules/stopping-microshift-executable.adoc:27]:
```suggestion
$ sudo systemctl stop microshift
```

File: modules/stopping-microshift-executable.adoc [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+----
++
+[NOTE]
+====
+If MicroShift is running as a service, you must execute the `sudo systemctl stop microshift` command to shut it down. You can review the output of the `journalctl -xu microshift` command to verify the service termination.
```
Comment (dhellmann) [inline modules/stopping-microshift-executable.adoc:32]:
```suggestion
You can review the output of the `journalctl -xu microshift` and `systemctl status microshift` commands to verify the service termination. 
```

File: modules/stopping-microshift-executable.adoc [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+====
+The full cleanup does not remove OVS configuration applies by the MicroShift service initialization sequence. Run the `sudo /usr/bin/configure-ovs.sh` command to revert to the original network settings. 
+====
+
+. Enter the following command to ensure that the MicroShift deployment has been successfully shut down:
+[source,terminal]
+----
+$ tail -3 /tmp/microshift.log 
+----
+
```
Comment (dhellmann) [inline modules/stopping-microshift-executable.adoc:52]:
```suggestion
```

These are dev instructions.

File: installing/installing_sno/installing-microshift-on-rhel.adoc [hunk 19]
```diff
@@ -0,0 +1,7 @@
+:_content-type: ASSEMBLY
+[id="installing-microshift-on-rhel"]
+= Installing MicroShift on {op-system-first}
+include::_attributes/common-attributes.adoc[]
+:context: microshift-installation-overview
+
+To install MicroShift in a production-like environment, you must create a {op-system} ISO installer that includes the necessary components pre-loaded on the image.
```
Comment (dhellmann) [inline installing/installing_sno/installing-microshift-on-rhel.adoc:7]:
```suggestion
To install MicroShift in a production-like environment, you will usually create a {op-system} ISO installer that includes the necessary components pre-loaded on the image. 
```

Having a system image is not an absolute requirement, but it's the most common way we expect for edge deployments.

File: modules/creating-microshift-development-environment.adoc [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+    --memory 4096 \
+    --disk path=./${VMNAME}.qcow2,size=50 \
+    --network network=default,model=virtio \
+    --os-type generic \
+    --events on_reboot=restart \
+    --cdrom ./rhel-8.6-$(uname -i)-boot.iso \
+"
+----
+
+. When prompted by the OS installation wizard, configure a MicroShift administrator user and root password.
```
Comment (dhellmann) [inline modules/creating-microshift-development-environment.adoc:54]: The actual user name does not matter here. We specified a user for the developer docs to have the dev environments consistent across the team, but we may not even need a user to create a separate administrative user if they don't want to.

File: modules/configuring-microshift-virtual-machine.adoc [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+$ sudo dnf update -y
+----
+
+. Run the following command to install basic dependencies: 
++
+[source,terminal]
+----
+$ sudo dnf install -y git cockpit make gcc selinux-policy-devel rpm-build bash-completion
+----
+
```
Comment (dhellmann) [inline modules/configuring-microshift-virtual-machine.adoc:39]:
```suggestion
```

@ggiguash do we really need cockpit, or is it just convenient? I never use it on my system. I'm fairly certain none of the other things are dependencies for just running MicroShift, so I think we can remove this step entirely unless we do need cockpit.

File: modules/configuring-microshift-virtual-machine.adoc [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+$ export GO_VER=1.18.7; curl -L -o go${GO_VER}.linux-amd64.tar.gz https://go.dev/dl/go${GO_VER}.linux-amd64.tar.gz &&
+    sudo rm -rf /usr/local/go${GO_VER} && \
+    sudo mkdir -p /usr/local/go${GO_VER} && \
+    sudo tar -C /usr/local/go${GO_VER} -xzf go${GO_VER}.linux-amd64.tar.gz --strip-components 1 && \
+    sudo rm -rfv /usr/local/bin/{go,gofmt}
+    sudo ln --symbolic /usr/local/go1.18.7/bin/go /usr/local/go1.18.7/bin/gofmt /usr/local/bin/ && \
+    rm -rfv go${GO_VER}.linux-amd64.tar.gz
+----
+
+. Access the VM cockpit by visiting `https://<vm_ip>:9090`.
```
Comment (dhellmann) [inline modules/configuring-microshift-virtual-machine.adoc:53]:
```suggestion
```

See the question about cockpit above.

**Comment by fzdarsky** [comment]
@stevsmit @dhellmann I'm a bit confused why this doc is based on https://github.com/openshift/microshift/blob/main/docs/devenv_rhel8.md, which is only for our engineers developing on MicroShift itself, *not for users*, so shouldn't be in the user documentation?

**Comment by stevsmit** [comment]
@fzdarsky yeah after writing this up and talking to Doug and Giga, it was then made clear that these wouldn't be the instructions we would be providing to customers.

**Pull request closed without merging**